Mathlib/Algebra/Homology/TotalComplexShift.lean

```diff
@@ -0,0 +1,209 @@
+import Mathlib.Algebra.Homology.TotalComplex
+
+/-!
+# Behaviour of the total complex with respect to shifts
+
+There are two ways to shift objects in `HomologicalComplex₂ C (up ℤ) (up ℤ)`:
+* by shifting the first indices (and changing signs of horizontal differentials),
+which corresponds to the shift by `ℤ` on `CochainComplex (CochainComplex C ℤ) ℤ`.
+* by shifting the second indices (and changing signs of vertical differentials).
+
+These two sorts of shift functors shall be abbreviated as
+`HomologicalComplex₂.shiftFunctor₁ C x` and
+`HomologicalComplex₂.shiftFunctor₂ C y`.
+
+In this file, for any `K : HomologicalComplex₂ C (up ℤ) (up ℤ)`, we define an isomorphism
+`K.totalShift₁Iso x : ((shiftFunctor₁ C x).obj K).total (up ℤ) ≅ (K.total (up ℤ))⟦x⟧`
+for any `x : ℤ` (which does not involve signs) and an isomorphism
+`K.totalShift₂Iso y : ((shiftFunctor₂ C y).obj K).total (up ℤ) ≅ (K.total (up ℤ))⟦y⟧`
+for any `y : ℤ` (which is given by the multiplication by `(p * y).negOnePow` on the
+summand in bidegree `(p, q)` of `K`).
+
+## TODO
+
+- show that the two sorts of shift functors on bicomplexes "commute", but that signs
+appear when we compare the compositions of the two compatibilities with the total complex functor.
+- deduce compatiblities with shifts on both variables of the action of a
+bifunctor on cochain complexes
+
+-/
+
+open CategoryTheory Category ComplexShape Limits
+
+namespace HomologicalComplex₂
+
+variable (C : Type*) [Category C] [Preadditive C]
+
+/-- The shift on bicomplexes obtained by shifting the first indices (and changing the
+sign of differentials). -/
+abbrev shiftFunctor₁ (x : ℤ) :
+    HomologicalComplex₂ C (up ℤ) (up ℤ) ⥤ HomologicalComplex₂ C (up ℤ) (up ℤ) :=
+  shiftFunctor _ x
+
+/-- The shift on bicomplexes obtained by shifting the second indices (and changing the
+sign of differentials). -/
+abbrev shiftFunctor₂ (y : ℤ) :
+    HomologicalComplex₂ C (up ℤ) (up ℤ) ⥤ HomologicalComplex₂ C (up ℤ) (up ℤ) :=
+  (shiftFunctor _ y).mapHomologicalComplex _
+
+variable {C}
+variable (K : HomologicalComplex₂ C (up ℤ) (up ℤ))
+
+section
+
+variable (x : ℤ) [K.HasTotal (up ℤ)] [((shiftFunctor₁ C x).obj K).HasTotal (up ℤ)]
+
+/-- Auxiliary definition for `totalShift₁Iso`. -/
+noncomputable def totalShift₁XIso (n n' : ℤ) (h : n + x = n') :
+    (((shiftFunctor₁ C x).obj K).total (up ℤ)).X n ≅ (K.total (up ℤ)).X n' where
+  hom := totalDesc _ (fun p q hpq => K.ιTotal (up ℤ) (p + x) q n' (by dsimp at hpq ⊢; omega))
+  inv := totalDesc _ (fun p q hpq =>
+    (K.XXIsoOfEq (Int.sub_add_cancel p x) rfl).inv ≫
+      ((shiftFunctor₁ C x).obj K).ιTotal (up ℤ) (p - x) q n
+        (by dsimp at hpq ⊢; omega))
+  hom_inv_id := by
+    ext p q h
+    dsimp
+    simp only [ι_totalDesc_assoc, CochainComplex.shiftFunctor_obj_X', ι_totalDesc, comp_id]
+    exact ((shiftFunctor₁ C x).obj K).XXIsoOfEq_inv_ιTotal _ (by omega) rfl _ _
+
+lemma D₁_totalShift₁XIso_hom (n₀ n₁ n₀' n₁' : ℤ) (h₀ : n₀ + x = n₀') (h₁ : n₁ + x = n₁') :
```

Comment:
I think this one and the lemma below are also missing `@[reassoc]`?

Reply:
Thanks for spotting this!